doc/design/service-discovery.md

```diff
@@ -0,0 +1,316 @@
+(saved locally) would be compared to the new service registry in a new class, extending the public user API.
+
+Pro:
+
+* Simple and consistent user API for both event-based and sychronous requests
+  * Filtering for `findService` could be done inside the new class
+* No need for a change counter as ABA problem solved by POSH mechanisms
+* Not just delta of service registry is transmitted, but complete service registry info
+* Fast data transport
+* Close to DDS mechanism
+* Could replace `InterfacePort`s
+
+Con:
+
+* New publisher in RouDi needed
+* Partially overlapping features with `InterfacePort`s
+  * Could the `InterfacePort`s be replace by alternative D?
+
+Note:
+
+* To avoid any out of memory issues with custom user-configured mempools or access rights problems, the publisher shall
+write into the `iceoryx_managment` segment (same applies for the introspection publishers)
+
+### Decision
+
+The **event-based notification alternative D** fulfills both requirements for a asychronous, one shot-polling approach as
+well as the event-based one. It can replace the `InterfacePort`s, which broadens the usage e.g. for gateways or even
+introspection 2.0. Furthermore, alternative D does not add any new non-trivial building blocks. The one-shot RPC call
+does not lead to a shared memory transmission compared to the alternative B (request/response feature usage) as the
+complete service registry is stored locally.
+
+### Code example
+
+#### Event-based notification: Alternative B
+
+```cpp
+auto& runtime = PoshRuntime::initRuntime("myApp");
+auto* interfacePortData = runtime.getMiddlewareInterface(capro::Interfaces::INTERNAL);
+
+InterfacePort caproSubscriber(interfacePortData);
+Listener myListner;
+
+void onDiscoveryUpdateCallback(InterfacePort* subscriber)
+{
+    subscriber->tryGetCaProMessage().and_then([](auto& caproMessage){
+        // Has any service that is relevant for me changed?
+        if(caproMessage.m_serviceDescription == myRelevantServiceDescription && someOtherCondition())
+        {
+            // call user-defined callback
+        }
+    });
+}
+
+myListner.attachEvent(caproSubscriber, DATA_RECEIVED, createNotificationCallback(onDiscoveryUpdateCallback));
+```
+
+#### Event-based notification: Alternative D
+
+```cpp
+class DiscoveryInfo
+{
+  public:
+
+    // Move all service-related methods from PoshRuntime to this class
+    cxx::expected<InstanceContainer, FindServiceError>
+    PoshRuntime::findService(const capro::ServiceDescription& serviceDescription) noexcept
+    {
+        // Update local service registry
+        get();
+
+        return filterFor(serviceDescription);
+    }
+
+    // @todo check if still needed
+    bool offerService(const capro::ServiceDescription& serviceDescription) noexcept;
+    void stopOfferService(const capro::ServiceDescription& serviceDescription) noexcept;
+
+    ServiceRegistryTopic get() noexcept
```

Comment:
As noted this also covers the one shot complete registry state update on demand.